### ISSUE TITLE
Using .mock() as a .default() only in testing?

### ISSUE THREAD

**Issue by agentlewis** [vriad/zod#195]
Hey hey,

So we are looking at using a "zod first approach" to developing libraries (DDD?).

As such we have just been looking at a way to quickly mock inputs during develop and testing. One idea we had was to use [.default()](https://github.com/vriad/zod/tree/v2#default-values) to produce the test data - but then we realised this didn't make sense as we would then have these defaults in production as well.

It would be great if there were an equivalent `.mock()` that would return a mock value if no other is received and node env == `dev` or `test` (or some other config).

cc @iainkirkpatrick

**Comment by lazharichir** [comment]
Hi @agentlewis, 

What I have done is I used `faker js` ([repo](https://github.com/marak/Faker.js/)). I have a function that takes a `ZodSchema` and goes through each of the `Schema.shape.KEY` and generates a fake value accordingly and then returns the object (parsing it first to be sure it is valid).

This is the create valid mocks, though (e.g. to fill a DB). To test wrong values, I do that manually generally.

**Comment by vriad** [comment]
I don't want to modify the behavior of `.default` in different environments. There are too many different ways of determining a dev/prod environment, and I don't want to be opinionated on that front. 

But I'm open to a PR for a separate `.mock` method. 

My preferred approach to this would be a `.mock` method on the base class. This method could accept some parameters that let you overriding the default mocking functions for each data type (e.g. strings, numbers, tuples, etc). 

There would also need to be a way to override the mocking function for a particular schema instance (probably a `.mocker()` or `.setMocker()` method on the base class). This lets you mock refinement types like emails or UUIDs. 

One complexity that comes to mind: 

- Refinements: Zod has no way of generating a value that satisfies a given refinement. There would also need to be a way to override the mocking function for a particular schema instance (probably a `.mocker()` or `.setMocker()` method on the base class). 
- Unions: which union "option" should be used to generate the value? Perhaps a randomly chosen one? Should the mocked value for an optional value be `undefined` 50% of the time?
- Transformers: should Zod mock a value for the input or the output type? Not a huge issue, this can just be an option/flag in the `.mock` method.

**Comment by agentlewis** [comment]
@lazharichir thanks :)

@vriad That looks pretty good. I agree that default should behave one way regardless of environment. So a seperate function `.mock()` that you would chain on was what I was thinking.

That solution of just adding a function to default looks like it takes care of my need.

**Comment by vriad** [comment]
I just opened an RFC for mocking here: https://github.com/vriad/zod/issues/197

Feel free to comment 🤙

**Comment by vriad** [comment]
The "dynamic default" (e.g. passing a function that returns a default value) is now available in zod@2.0.0-beta.16 👍

```ts
const randomDefault = z
    .number()
    .default(() => Math.random())

// returns a new random number every time `parse` is called
randomDefault.parse(undefined);
```

Zod makes sure the returned value matches the expected type signature:

```ts
const randomDefault = z
    .number()
    .default(() => 'qwerty') // TypeError
```
